### DISCUSSION TITLE
How to harvest with CKAN 2.9 ?

### DISCUSSION THREAD

**Discussion by SorrowAvailable** [ckan/ckan#6859] [Q&A]
Hello, I am in charge to evaluate the potential of CKAN without any positive outcome so far when it comes to harvesting despite having the proper install including /harvest link. Are there people able to describe the way I can follow to have a working harvesting campaign ? Regards,

**Comment by duttonw** [comment]
Hi @SorrowAvailable ,

I work for the Queensland Government and per our policy of being open, i can share these details with you today.

Our production instance is on 2.8.x and will be on 2.9.x in the coming weeks. We will be getting our staging environment onto 2.9.x next week.

We are also in the process of preparing to harvesting a large instance into www.data.qld.gov.au. It is currently in our dev region and is on ckan 2.9.

Here's how we harvested geoscience ckan instance into our own. note that we have some customisations done to not take in resources due to also including xloader and archiver which would have cloned the content into our database and storage device and said ckan instance points to terabytes of data. As we needed to match schemas we have an extension on top of the original harvester plugin to accommodate this. https://github.com/qld-gov-au/ckanext-harvester-data-qld-geoscience/

You will need to ensure you have the harvest job workers running in the background. Per harvester setup, as well as the cron job's to fire off any scheduled jobs and job closures once complete.


pre-req to harvesting is to ensure you have an organisation created that it will be attaching said resources to.

* go to ${ckan site}/harvest/new
* give it the full url to the site you will be harvesting i.e. https://geoscience.data.qld.gov.au/
* give the harvester job a name and a nice description (non logged in users can see this data)
* select your 'source type' normally that would be ckan if they are vanilla, or your own custom schema if loaded. for us this was 'GSQ Open Data Portal'
* pick how often and now the tricky part is the configuration for the job.

  for our purposes we wanted to overwrite certain fields with defaults so our configuration looked like this:
  ```
   {"default_groups": ["geoscience"], 
   "default_group_dicts": [{"display_name": "Geoscience", "description": "", "image_display_url": "", 
   "package_count": 0, 
   "created": "2022-03-24T12:53:07.848442", 
   "name": "geoscience", "is_organization": false, "state": "active", "extras": [], "image_url": "", 
   "groups": [], "type": "group", "title": "Geoscience",
    "revision_id": "7e80657d-3a57-4364-92d0-95125f5e4cae", "num_followers": 0, 
   "id": "eea6716f-b97e-4f07-ad5f-1edf68d8a373", 
   "tags": [], "approval_status": "approved"}], "version": "1.0", 
   "deletion_reason": "Dataset deleted at harvest source", 
   "author_email": "gsqopendata@resources.qld.gov.au", "de_identified_data": "NO", 
  "data_driven_application": "NO", "dataset_type":    "geoscience", "update_frequency": "non-regular", "security_classification": "PUBLIC", "license_id": "cc-by-4"}
   ```
   note that our base schema has lots of extra fields from customisations, your's would be far simpler. 

once the job is kicked off and the gather worker picked up the task, it will scan the remote ckan instance for all datasets and generate tasks for each. 

For this ckan instance of 170k datasets, this took over 15 minutes to complete. Then another 2 days of straight importing to get all details in.

You can find our plugin list and cook book here https://github.com/qld-gov-au/ckan-qld-infrastructure/blob/master/vars/shared-OpenData.var.yml and here https://github.com/qld-gov-au/opswx-ckan-cookbook.

Do note that we have forked most plugins to make 2.9.x compatible and are slowly providing this alterations back upstream (minus the additional tweaks our business owners have requested).

Here is our plugin list https://www.data.qld.gov.au/api/action/status_show at this time it looks like this: 
```
{"help": "https://www.data.qld.gov.au/api/3/action/help_show?name=status_show", "success": true, "result": {"site_url": 
"https://www.data.qld.gov.au", "site_description": "", "extensions": ["stats", "resource_proxy", "text_preview", "recline_preview", 
"text_view", "webpage_view", "recline_grid_view", "image_view", "recline_view", "recline_graph_view", "recline_map_view", 
"odi_certificates", "dcat", "structured_data", "data_qld_resources", "data_qld_integration", "data_qld_google_analytics", 
"data_qld_reporting", "scheming_datasets", "qa", "archiver", "csrf_filter", "datarequests", "harvest", "ckan_harvester", "report", 
"harvester_data_qld_geoscience", "qgovext", "resource_type_validation", "s3filestore", "ssm_config", "validation", "xloader", 
"ytp_comments", "datastore"], "error_emails_to": null, "locale_default": "en_AU", "site_title": "Open Data Portal | Queensland Government"}}
```


Regards

@duttonw

**Reply by SorrowAvailable** [reply]
Hello and Thank-You for your detailed reply,
As a matter of fact, I had found, scattered on internet, elements of information leading me to add the following elements in the configuration section of the harvesting :
{
 "api_version": 1,
 "default_tags": [{"name": "geo"}, {"name": "namibia"}],
 "default_groups": ["science", "spend-data"],
 "default_extras": {"encoding":"utf8", "harvest_url": "{harvest_source_url}/dataset/{dataset_id}"},
 "override_extras": true,
 "organizations_filter_exclude": ["remote-organization"],
 "user":"me",
 "read_only": true,
 "remote_groups": "only_local",
 "remote_orgs": "create"
}
despite the harvesting mode to be daily, I have never seen any other status but "running" regarding the harvesting process while I was expected some "Started" or any kind of error message. The consequence being to debug with command line :
$ ckan --config=/etc/ckan/default/ckan.ini harvester run-test
Showing me :
Usage: ckan harvester run-test [OPTIONS] SOURCE_ID_OR_NAME GUID

I tried any replacement for both SOURCE_ID_OR_NAME and GUID without any success except python exception stack trace not giving me direction to where expected elements are to be found in the CKAN console !
This worst case OpenSource usage scenario suffers from the poorness of documentation and the lack of precision on step-by-step quality control on the whole process to ensure the predating steps are fulfilled before jumping forward.
Thank-You for sharing your experience on this matter and I hope positive outcome for both your migration and my attempt to use the product.
Do you have a minimalist harvesting configuration that could fit testing the harvesting process ?
Regards,

**Reply by SorrowAvailable** [reply]
Hello and sorry to disturb you again but as you can see on the following screenshots, I am still not able to perform a single harvesting and I raise concern on the configuration because, from the example you provided above the instruction :
 {"default_groups": ["geoscience"], 
Generates the message : Error parsing the configuration options: Default group not found
that forces me to remove the instruction from the configuration for the harvest to be saved but still, without starting !

![Screenshot 2022-05-16 at 18-07-19 harvester demo 1 - CKAN](https://user-images.githubusercontent.com/105364033/168636624-8b892e8a-8f1e-4492-a5d6-448c2113b020.png)
![Screenshot 2022-05-16 at 18-06-26 Edit harvest source - CKAN](https://user-images.githubusercontent.com/105364033/168636629-78a2b41f-915a-48fa-9a8f-79db26db9543.png)

**Reply by duttonw** [reply]
for us
```${ckanurl}/group/geoscience```
if not exists
```${ckanurl}/group/new```

and create it.

tail the harvester gather worker logs to know what is occurring.

you can always hit cancel after a while to see if there was any errors etc.

**Reply by duttonw** [reply]
i've done more testing for you. note this is with our custom ckan site that has a large number of extra features so i used our custom schema (which also adds default fields to pass our custom validators)


after all 118 datasets were ingested, the hourly cron job needs to be run to 'complete' the job.

PASTER_PLUGIN=ckanext-harvest /usr/lib/ckan/default/bin/ckan_cli harvester run
which for 2.9 turns into 
```exec /usr/lib/ckan/default/bin/ckan -c /etc/ckan/default/production.ini harvester run```

you can see it here:
https://dev.data.qld.gov.au/group/birmingham
https://dev.data.qld.gov.au/harvest/birmingham-data-test

config
```
{"default_groups": ["birmingham"], "deletion_reason": "Dataset deleted at harvest source", "default_group_dicts": [{"display_name": "birmingham", "description": "test birmingham", "image_display_url": "", "package_count": 0, "created": "2022-05-17T11:11:20.023075", "name": "birmingham", "is_organization": false, "state": "active", "extras": [], "image_url": "", "groups": [], "type": "group", "title": "birmingham", "num_followers": 0, "id": "5e6ff3f6-d0dd-4a85-9f2f-b52b40b57c9b", "tags": [], "approval_status": "approved"}], "author_email": "example@example.com", "data_driven_application": "NO", "version": "1.0", "de_identified_data": "NO", "security_classification": "PUBLIC", "license_id": "cc-by-4", "update_frequency": "non-regular", "dataset_type": "ckan"}
```

detailed job report
Job Report
1 errors 117 added 0 updated 0 deleted 0 not modified

Id	468faee6-9713-4008-bfff-5c2bf91661c6
Created	17 May 2022, 11:20 AM (UTC+10:00)
Started	17 May 2022, 11:20 AM (UTC+10:00)
Finished	17 May 2022, 11:21 AM (UTC+10:00)
Status	Finished
Error Summary Only the 20 most frequent errors are shown
Document Errors
Count	Message
1	Invalid package with GUID cf552d08-cee9-43bf-8c0f-3196a9311799: {'maintainer_email': [u'Email ian.martin@birmingham.gov.uk is not a valid format']}

**Reply by duttonw** [reply]
I should mention that that dev environment is only up for AEST business hours (GMT+10) . So it may be offline to you depending on what time you check it

**Comment by SorrowAvailable** [comment]
Hello William,
My attempt to follow your instruction is not giving me result where the harvesting command is showing :
![image](https://user-images.githubusercontent.com/105364033/168810150-998a9ccb-4508-4ee2-8a08-9e105d30e74c.png)

While I cannot see any result from the CKAN interface. 
How could I abuse a dozen minutes of your time on Friday the 20th @ 1000 for you and midnight for me ? I would like to have a Teams meeting and to show you my screen for you to pinpoint discrepancies I am facing compared to the harvesting expectations. What means do I have to send you a Teams invitation ?
Kind Regards,

**Reply by duttonw** [reply]
Hi @SorrowAvailable , you do not have the gather and fetch workers enabled, if you look at your queue they are not on the default queue and not picked up. this is also the same if you install the archiver plugin, you are required to setup priority and bulk queue listeners. I am unsure if you can make it listen to multiple queues.

see our config here
https://github.com/qld-gov-au/opswx-ckan-cookbook/blob/master/files/default/supervisor-ckan-harvest.conf

and
https://github.com/qld-gov-au/opswx-ckan-cookbook/blob/master/recipes/ckanweb-deploy-exts.rb#L205
https://github.com/qld-gov-au/opswx-ckan-cookbook/blob/5f82f8fc8959ed996b8544c01ffc4796c5b736c9/recipes/ckanbatch-setup.rb

**Comment by duttonw** [comment]
Hi @SorrowAvailable , did this solve your problem?

**Reply by duttonw** [reply]
I believe this has now been answered since no response has been received from @SorrowAvailable  in the last 7 days.

**Comment by SorrowAvailable** [comment]
Hello,
The subject is far from being answered since I still cannot havest and I am at the same point with no result so far. The subject will be addressed when I will be able to have a working harvesting process that will lead me to claim victory on this very page for any others to follow the same path.
Could you tell me what output I could bring to your knowledge for identifying what element is disfuctioning ?
On my end, I have to report that the material you supply me is absolutely incompatible with the install I have. Your config file describes :
command=/usr/lib/ckan/default/bin/ckan_cli harvester fetch_consumer
That does not exist on my end, rather I have to :
command=/usr/lib/ckan/default/bin/ckan --config=/etc/ckan/default/ckan.ini harvester fetch_consumer for having a launching command, still, without harvesting because the command is telling me that a harvesting process has already started ; which is true, the CKAN GUI has started a harvesting process which does not harvest at all despite the daily or manual mode and restarting the service each time a modification is don e to the configuration .
Thand-You && Regard,
